src/drivers/uavcan_v1/Uavcan.hpp

```diff
@@ -0,0 +1,171 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <px4_platform_common/px4_config.h>
+#include <px4_platform_common/atomic.h>
+#include <px4_platform_common/defines.h>
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/module_params.h>
+#include <px4_platform_common/px4_work_queue/ScheduledWorkItem.hpp>
+
+#include <lib/parameters/param.h>
+#include <lib/perf/perf_counter.h>
+#include <uORB/Publication.hpp>
+#include <uORB/Subscription.hpp>
+#include <uORB/SubscriptionCallback.hpp>
+#include <uORB/topics/battery_status.h>
+#include <uORB/topics/parameter_update.h>
+#include <uORB/topics/sensor_gps.h>
+
+#include "o1heap/o1heap.h"
+
+#include <canard.h>
+#include <canard_dsdl.h>
+
+#include <reg/drone/srv/battery/Status_0_1.h>
+#include <reg/drone/srv/battery/Parameters_0_1.h>
+#include <uavcan/node/Heartbeat_1_0.h>
+
+//Quick and Dirty PNP imlementation only V1 for now as well
+#include <uavcan/node/ID_1_0.h>
+#include <uavcan/pnp/NodeIDAllocationData_1_0.h>
+#include <uavcan/pnp/NodeIDAllocationData_2_0.h>
+
+//Quick and Dirty UAVCAN register implementation
+#include <uavcan/_register/List_1_0.h>
+#include <uavcan/_register/Access_1_0.h>
+
+#define PNP1_PORT_ID                                 uavcan_pnp_NodeIDAllocationData_1_0_FIXED_PORT_ID_
+#define PNP1_PAYLOAD_SIZE                            uavcan_pnp_NodeIDAllocationData_1_0_SERIALIZATION_BUFFER_SIZE_BYTES_
+#define PNP2_PORT_ID                                 uavcan_pnp_NodeIDAllocationData_2_0_FIXED_PORT_ID_
+#define PNP2_PAYLOAD_SIZE                            uavcan_pnp_NodeIDAllocationData_2_0_SERIALIZATION_BUFFER_SIZE_BYTES_
+
+#include "CanardInterface.hpp"
+
+class UavcanNode : public ModuleParams, public px4::ScheduledWorkItem
+{
+	/*
+	 * This memory is reserved for uavcan to use as over flow for message
+	 * Coming from multiple sources that my not be considered at development
+	 * time.
+	 *
+	 * The call to getNumFreeBlocks will tell how many blocks there are
+	 * free -and multiply it times getBlockSize to get the number of bytes
+	 *
+	 */
+	static constexpr unsigned HeapSize = 8192;
```

Comment:
The comment provided for this entry is invalid because libcanard does not use block allocators.

The amount of memory provided here is likely to be insufficient even for basic PX4 applications due to fragmentation. If memory size is a concern, it might make sense to use the default OS heap instead of O1Heap, although I am not sure if it meets the temporal requirements.